src/DataProtection/DataProtection/test/HostingTests.cs

```diff
@@ -46,6 +47,32 @@ public async Task LoadsKeyRingBeforeServerStarts()
             mockKeyRing.VerifyAll();
         }
 
+        [Fact]
+        public async Task GenericHostLoadsKeyRingBeforeServerStarts()
+        {
+            var tcs = new TaskCompletionSource<object>();
+            var mockKeyRing = new Mock<IKeyRingProvider>();
+            mockKeyRing.Setup(m => m.GetCurrentKeyRing())
+                .Returns(Mock.Of<IKeyRing>())
+                .Callback(() => tcs.TrySetResult(null));
+
+            var builder = new HostBuilder()
+                .ConfigureServices(s =>
+                    s.AddDataProtection()
+                    .Services
+                    .Replace(ServiceDescriptor.Singleton(mockKeyRing.Object))
+                    .AddSingleton<IServer>(
+                        new FakeServer(onStart: () => tcs.TrySetException(new InvalidOperationException("Server was started before key ring was initialized")))));
```

Comment:
This IServer isn't actually getting started, you didn't add the webhostedservice.

Reply:
You mean adding a `.ConfigureWebHost(b => b.UseStartup<TestStartup>());` right?